### PR TITLE
feat: Add Challenge headline and description

### LINKED ISSUE #1119: Add Challenge tagline
The Challenge card includes a tagline in the challenge registry mockup. There are two strategy to capture this information:

- Add a dedicated property, e.g. `Challenge.tagline`
- Use the property `Challenge.description`

For context, I'm currently working on text search with Hibernate Search 6 and Elasticsearch. This work reminds me of the importance of accepting large descriptions (`Challenge.description`) in order to increase the relevance of text search results.

But there is still value in having a shorter tagline that fits into the challenge card. Hence this proposal:

1. If `Challenge.tagline` is specified, we use it in the Challenge card,
2. otherwise, we display the first N characters from `Challenge.description`.

That way, we don't absolutely need to come up with a tagline for all challenges, though this a tagline is probably a plus to capture the attention of the users.

How does this proposal sound? @vpchung @rrchai

### LINKED ISSUE #1120: Add property `Challenge.description`
Need at this point to test text search using multiple fields.

### PR DESCRIPTION
**Pull request by tschaffter**
- Fixes #1119 
- Fixes #1120 

## Changelog

- Add the property `Challenge.headline`
  - The recommended length in this article is 60 chars. Here I give a bit more room and set the max length to 80 chars.
- Add the property `Challenge.description`
  - The recommended length in this article is between 150 and 160 characters. Here I set it to 280 (tweet length).
  - 280 chars is a compromise between a suitable description for Google SEO and a longer description that contribute to the relevance of results when using text search.
  - Happy to revisit this decision if additional considerations are available.

## Preview

List the challenges:

```console
GET {{basePath}}/challenges

{
  "number": 0,
  "size": 100,
  "totalElements": 3,
  "totalPages": 1,
  "hasNext": false,
  "hasPrevious": false,
  "challenges": [
    {
      "id": 1,
      "name": "The Digital Mammography DREAM Challenge",
      "headline": "Example headline",
      "description": "Example description",
      "status": "upcoming",
      "difficulty": "good_for_beginners",
      "platform": {
        "id": 1,
        "name": "synapse",
        "displayName": "Synapse"
      },
      "incentives": [
        "monetary",
        "publication"
      ],
      "submissionTypes": [
        "container_image",
        "prediction_file"
      ],
      "createdAt": "2023-01-04T23:56:20Z",
      "updatedAt": "2023-01-04T23:56:20Z"
    },
    {
      "id": 2,
      "name": "Patient Mortality EHR DREAM Challenge",
      "headline": "Example headline",
      "description": "Example description",
      "status": "active",
      "difficulty": "intermediate",
      "platform": {
        "id": 1,
        "name": "synapse",
        "displayName": "Synapse"
      },
      "incentives": [
        "monetary"
      ],
      "submissionTypes": [
        "container_image"
      ],
      "createdAt": "2023-01-04T23:56:20Z",
      "updatedAt": "2023-01-04T23:56:20Z"
    },
    {
      "id": 3,
      "name": "COVID-19 EHR DREAM Challenge",
      "headline": "Example headline",
      "description": "Example description",
      "status": "completed",
      "difficulty": "advanced",
      "platform": {
        "id": 2,
        "name": "codalab",
        "displayName": "CodaLab"
      },
      "incentives": [],
      "submissionTypes": [],
      "createdAt": "2023-01-04T23:56:20Z",
      "updatedAt": "2023-01-04T23:56:20Z"
    }
  ]
}
```

### CODE REVIEW

File: apps/challenge-registry/challenge-service/src/main/java/org/sagebionetworks/challenge/model/dto/ChallengeDto.java [hunk 1]
```diff
@@ -101,6 +107,55 @@ public void setName(String name) {
...
+   * The headline of the a challenge
+   *
+   * @return description
+   */
+  @NotNull
+  @Size(min = 0, max = 280)
+  @Schema(
+      name = "description",
+      example = "This is an example challenge description.",
+      description = "The headline of the a challenge",
```
Comment (rrchai) [inline apps/challenge-registry/challenge-service/src/main/java/org/sagebionetworks/challenge/model/dto/ChallengeDto.java:149]: The description of a challenge?

File: apps/challenge-registry/challenge-service/src/main/java/org/sagebionetworks/challenge/model/dto/ChallengeDto.java [hunk 2]
```diff
@@ -101,6 +107,55 @@ public void setName(String name) {
...
+   *
+   * @return headline
+   */
+  @Size(min = 0, max = 80)
+  @Schema(
+      name = "headline",
+      example = "Example challenge headline",
+      description = "The headline of the a challenge",
+      required = false)
+  public String getHeadline() {
```
Comment (vpchung) [inline apps/challenge-registry/challenge-service/src/main/java/org/sagebionetworks/challenge/model/dto/ChallengeDto.java:126]: Ah, back to the good ol' getters and setters -- it's been awhile!

File: apps/challenge-registry/challenge-service/src/main/java/org/sagebionetworks/challenge/model/dto/ChallengeDto.java (hunk 1 above)
Comment (vpchung) [inline apps/challenge-registry/challenge-service/src/main/java/org/sagebionetworks/challenge/model/dto/ChallengeDto.java:149]:
```suggestion
      description = "The description of a challenge",
```

File: libs/challenge-registry/api-description/build/challenge.openapi.yaml [hunk 3]
```diff
@@ -249,6 +249,18 @@ components:
...
       pattern: ^[a-z0-9]+(?:-[a-z0-9]+)*$
       example: awesome-challenge
+    ChallengeHeadline:
+      description: The headline of the a challenge
+      type: string
+      minLength: 0
+      maxLength: 80
+      example: Example challenge headline
+    ChallengeDescription:
+      description: The headline of the a challenge
```
Comment (rrchai) [inline libs/challenge-registry/api-description/build/challenge.openapi.yaml:259]: The description of a challenge?

File: apps/challenge-registry/challenge-service/src/main/java/org/sagebionetworks/challenge/model/dto/ChallengeDto.java [hunk 4]
```diff
@@ -101,6 +107,55 @@ public void setName(String name) {
...
+  /**
+   * The headline of the a challenge
+   *
+   * @return headline
+   */
+  @Size(min = 0, max = 80)
+  @Schema(
+      name = "headline",
+      example = "Example challenge headline",
+      description = "The headline of the a challenge",
```
Comment (vpchung) [inline apps/challenge-registry/challenge-service/src/main/java/org/sagebionetworks/challenge/model/dto/ChallengeDto.java:124]:
```suggestion
      description = "The headline of a challenge",
```

File: libs/challenge-registry/api-description/build/openapi.yaml [hunk 5]
```diff
@@ -307,6 +307,18 @@ components:
...
       pattern: '^[a-z0-9]+(?:-[a-z0-9]+)*$'
       example: awesome-challenge
+    ChallengeHeadline:
+      description: The headline of the a challenge
+      type: string
+      minLength: 0
+      maxLength: 80
+      example: Example challenge headline
+    ChallengeDescription:
+      description: The headline of the a challenge
```
Comment (rrchai) [inline libs/challenge-registry/api-description/build/openapi.yaml:317]: same question above

File: libs/challenge-registry/api-description/build/openapi.yaml [hunk 6]
```diff
@@ -307,6 +307,18 @@ components:
       maxLength: 60
       pattern: '^[a-z0-9]+(?:-[a-z0-9]+)*$'
       example: awesome-challenge
+    ChallengeHeadline:
+      description: The headline of the a challenge
```
Comment (rrchai) [inline libs/challenge-registry/api-description/build/openapi.yaml:311]: same typo

File: apps/challenge-registry/challenge-service/src/main/java/org/sagebionetworks/challenge/model/dto/ChallengeDto.java (hunk 4 above)
Comment (rrchai) [inline apps/challenge-registry/challenge-service/src/main/java/org/sagebionetworks/challenge/model/dto/ChallengeDto.java:124]: same typo

File: libs/challenge-registry/api-description/build/openapi.yaml (hunk 5 above)
Comment (vpchung) [inline libs/challenge-registry/api-description/build/openapi.yaml:317]:
```suggestion
      description: The description of a challenge
```

File: libs/challenge-registry/api-description/build/openapi.yaml (hunk 6 above)
Comment (vpchung) [inline libs/challenge-registry/api-description/build/openapi.yaml:311]:
```suggestion
      description: The headline of a challenge
```

File: libs/challenge-registry/api-description/build/challenge.openapi.yaml [hunk 7]
```diff
@@ -249,6 +249,18 @@ components:
       maxLength: 60
       pattern: ^[a-z0-9]+(?:-[a-z0-9]+)*$
       example: awesome-challenge
+    ChallengeHeadline:
+      description: The headline of the a challenge
```
Comment (rrchai) [inline libs/challenge-registry/api-description/build/challenge.openapi.yaml:253]: same typo

File: libs/challenge-registry/api-description/src/components/schemas/ChallengeDescription.yaml [hunk 8]
```diff
@@ -0,0 +1,5 @@
+description: The headline of the a challenge
```
Comment (vpchung) [inline libs/challenge-registry/api-description/src/components/schemas/ChallengeDescription.yaml:1]:
```suggestion
description: The description of a challenge
```

File: libs/challenge-registry/api-description/src/components/schemas/ChallengeHeadline.yaml [hunk 9]
```diff
@@ -0,0 +1,5 @@
+description: The headline of the a challenge
```
Comment (vpchung) [inline libs/challenge-registry/api-description/src/components/schemas/ChallengeHeadline.yaml:1]:
```suggestion
description: The headline of a challenge
```

**Review by vpchung — CHANGES_REQUESTED** [review]

File: libs/challenge-registry/api-description/src/components/schemas/ChallengeDescription.yaml (hunk 8 above)
Comment (rrchai) [inline libs/challenge-registry/api-description/src/components/schemas/ChallengeDescription.yaml:1]: same typo

File: libs/challenge-registry/api-description/src/components/schemas/ChallengeHeadline.yaml (hunk 9 above)
Comment (rrchai) [inline libs/challenge-registry/api-description/src/components/schemas/ChallengeHeadline.yaml:1]: same typo

**Comment by vpchung** [comment]
Minor typos, but otherwise, LGTM.

**Comment by vpchung** [comment]
Ah, looks like @rrchai caught the same typos ;-)

**Review by rrchai — CHANGES_REQUESTED** [review]
Thanks. Just some typos.

**Comment by rrchai** [comment]
> Ah, looks like @rrchai caught the same typos ;-)

@vpchung Ah, haha. You are faster!

File: apps/challenge-registry/challenge-service/src/main/java/org/sagebionetworks/challenge/model/dto/ChallengeDto.java (hunk 1 above)
Comment (tschaffter) [reply apps/challenge-registry/challenge-service/src/main/java/org/sagebionetworks/challenge/model/dto/ChallengeDto.java]: Fixed

File: apps/challenge-registry/challenge-service/src/main/java/org/sagebionetworks/challenge/model/dto/ChallengeDto.java (hunk 2 above)
Comment (tschaffter) [reply apps/challenge-registry/challenge-service/src/main/java/org/sagebionetworks/challenge/model/dto/ChallengeDto.java:126]: They are even better when they are automatically generated like here!

File: libs/challenge-registry/api-description/build/challenge.openapi.yaml (hunk 3 above)
Comment (tschaffter) [reply libs/challenge-registry/api-description/build/challenge.openapi.yaml]: Fixed

File: libs/challenge-registry/api-description/build/openapi.yaml (hunk 5 above)
Comment (tschaffter) [reply libs/challenge-registry/api-description/build/openapi.yaml]: Fixed

File: libs/challenge-registry/api-description/build/openapi.yaml (hunk 6 above)
Comment (tschaffter) [reply libs/challenge-registry/api-description/build/openapi.yaml]: Fixed

File: apps/challenge-registry/challenge-service/src/main/java/org/sagebionetworks/challenge/model/dto/ChallengeDto.java (hunk 4 above)
Comment (tschaffter) [reply apps/challenge-registry/challenge-service/src/main/java/org/sagebionetworks/challenge/model/dto/ChallengeDto.java]: Fixed

File: libs/challenge-registry/api-description/build/challenge.openapi.yaml (hunk 7 above)
Comment (tschaffter) [reply libs/challenge-registry/api-description/build/challenge.openapi.yaml]: Fixed

File: apps/challenge-registry/challenge-service/src/main/java/org/sagebionetworks/challenge/model/dto/ChallengeDto.java (hunk 1 above)
Comment (tschaffter) [reply apps/challenge-registry/challenge-service/src/main/java/org/sagebionetworks/challenge/model/dto/ChallengeDto.java]: Fixed

File: apps/challenge-registry/challenge-service/src/main/java/org/sagebionetworks/challenge/model/dto/ChallengeDto.java (hunk 4 above)
Comment (tschaffter) [reply apps/challenge-registry/challenge-service/src/main/java/org/sagebionetworks/challenge/model/dto/ChallengeDto.java]: Fixed

File: libs/challenge-registry/api-description/build/openapi.yaml (hunk 5 above)
Comment (tschaffter) [reply libs/challenge-registry/api-description/build/openapi.yaml]: Fixed

File: libs/challenge-registry/api-description/build/openapi.yaml (hunk 6 above)
Comment (tschaffter) [reply libs/challenge-registry/api-description/build/openapi.yaml]: Fixed

File: libs/challenge-registry/api-description/src/components/schemas/ChallengeDescription.yaml (hunk 8 above)
Comment (tschaffter) [reply libs/challenge-registry/api-description/src/components/schemas/ChallengeDescription.yaml]: Fixed

File: libs/challenge-registry/api-description/src/components/schemas/ChallengeHeadline.yaml (hunk 9 above)
Comment (tschaffter) [reply libs/challenge-registry/api-description/src/components/schemas/ChallengeHeadline.yaml]: Fixed

File: libs/challenge-registry/api-description/src/components/schemas/ChallengeDescription.yaml (hunk 8 above)
Comment (tschaffter) [reply libs/challenge-registry/api-description/src/components/schemas/ChallengeDescription.yaml]: Fixed

File: libs/challenge-registry/api-description/src/components/schemas/ChallengeHeadline.yaml (hunk 9 above)
Comment (tschaffter) [reply libs/challenge-registry/api-description/src/components/schemas/ChallengeHeadline.yaml]: Fixed

**Comment by tschaffter** [comment]
@rrchai @vpchung Thanks for spotting the typos!

The reason of the large number of typo occurrences is that there are quite a few OpenAPI description files around. In the future, only the files includes in the `src/` folder of the `challenge-registry-api-description` project should be reviewed. The other files are automatically generated. I realize it's not ideal when it comes to PR reviews. Hopefully the file menu under "Files changed" can help spot the OpenAPI description files to review.

image

**Pull request merged**